TalkSDKSamples/TalkSDKSamples-ObjectiveC/ZENAllCustomViewController.m

```diff
@@ -0,0 +1,312 @@
+{
+    if (self.talkCall != nil) {
+        [self.talkCall disconnect];
+        self.talkCall = nil;
+    }
+
+    ZDKTalkCallData *callData = [[ZDKTalkCallData alloc] initWithDigitalLine:ZENZendeskDigitalLine
+                                                      recordingConsentAnswer:answer];
+    __weak typeof(self) weakSelf = self;
+    self.talkCall = [self.talk callWithCallData:callData statusChangeHandler:^(enum CallStatus status, NSError *error) {
+        [weakSelf onCallStatusChange:status error:error];
+    }];
+
+    self. callDurationLabel.text = @"00:00";
+}
+
+- (void)onCallStatusChange:(CallStatus)status error:(NSError *)error
+{
+    switch (status) {
+        case CallStatusConnecting:
+            self.agentAvailabilityLabel.text = @"Connecting...";
+            self.agentAvailabilityLabel.textColor = [UIColor darkGrayColor];
+            self.ongoingCallView.hidden = NO;
+            self.callButton.enabled = NO;
+            [self setOngoingCallControlsEnabled:NO];
+            [self startTimer];
+            break;
+        case CallStatusConnected:
+            self.agentAvailabilityLabel.text = @"";
+            self.ongoingCallView.hidden = NO;
+            self.callButton.enabled = NO;
+            [self setOngoingCallControlsEnabled:YES];
+            break;
+        case CallStatusDisconnected:
+            self.agentAvailabilityLabel.text = @"";
+            self.ongoingCallView.hidden = NO;
+            self.callButton.enabled = YES;
+            [self disableOngoingCallControlsAndHideWithDelay];
+            [self stopTimer];
+            break;
+        case CallStatusFailed:
+            self.agentAvailabilityLabel.text = @"An error occured. Please try again later";
+            self.agentAvailabilityLabel.textColor = [UIColor redColor];
+            self.ongoingCallView.hidden = NO;
+            self.callButton.enabled = YES;
+            [self disableOngoingCallControlsAndHideWithDelay];
+            [self stopTimer];
+            break;
+        default:
+            break;
+    }
+}
+
+- (void)disableOngoingCallControlsAndHideWithDelay
+{
+    [self setOngoingCallControlsEnabled:NO];
+
+    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(2 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
+        self.ongoingCallView.hidden = YES;
+    });
+}
+
+- (void)setOngoingCallControlsEnabled:(BOOL)enabled
+{
+    self.speakerSwitchLabel.textColor = enabled ? [UIColor blackColor] : [UIColor lightGrayColor];
+    self.speakerSwitch.enabled = enabled;
+    self.muteSwitchLabel.textColor = enabled ? [UIColor blackColor] : [UIColor lightGrayColor];
+    self.muteSwitch.enabled = enabled;
+    self.endCallButton.enabled = enabled;
+}
+
+#pragma mark - Call duration timer
+
+- (void)startTimer
+{
+    __weak typeof(self) weakSelf = self;
+    self.callTimer = [NSTimer scheduledTimerWithTimeInterval:0.25 repeats:YES block:^(NSTimer * _Nonnull timer) {
```

Comment:
This timer duration is reused from our implementation. Basically on a tick we are reading the duration and updating the label, since the time intervals are floating point numbers having a refresh time interval of 1.0 was prone to glitches when updating the label.